File: firmware/core/market.cpp
#include "market.hpp"
#include <algorithm>
#include <climits>
namespace bm {
const char *error_text(Error e) {
  static const char *const names[] = {
      "TRADE CONFIRMED",      "MARKET FULL",       "UNKNOWN PLAYER",
      "SYNC REQUIRED",        "NOT ENOUGH SOL",    "INVALID AMOUNT",
      "4 HOLDINGS MAX",       "NOT ENOUGH TOKENS", "COIN CLOSED",
      "CREATOR ONLY",         "RUG COOLDOWN",      "ONE ACTIVE COIN",
      "SYMBOL TAKEN/INVALID", "WAIT 20 SECONDS",   "INVALID GAME PROOF",
      "EPOCH REWARD CAP",     "SESSION EXHAUSTED"};
  auto i = unsigned(e);
  return i < 17 ? names[i] : "UNKNOWN ERROR";
}
uint32_t crc32(const uint8_t *s, size_t n) {
  uint32_t c = ~0u;
  while (n--) {
    c ^= *s++;
    for (int i = 0; i < 8; i++)
      c = (c >> 1) ^ (0xedb88320u & uint32_t(-int(c & 1)));
  }
  return ~c;
}
int64_t cost(int q, int n) {
  if (q < 0 || n < 1 || n > 10000 || q + n > 10000)
    return -1;
  return int64_t(n) * 100 + 5LL * n * (2 * q + n - 1) / 2;
}
uint8_t command(uint32_t s, unsigned r, unsigned &d) {
  for (unsigned i = 0; i < r; i++)
    s = 1664525u * s + 1013904223u;
  d = 400 + (s % 5) * 100;
  return (s >> 16) % 3;
}
uint32_t request_hash(const Request &r) {
  uint8_t b[26];
  Writer w{b, sizeof b};
  w.u8(uint8_t(r.op));
  w.u8(r.coin);
  w.u16(r.amount);
  w.bytes(r.symbol, 6);
  w.bytes(r.proof, 16);
  return crc32(b, w.pos);
}
static Holding *holding(Player &p, unsigned coin) {
  Holding *empty = nullptr;
  for (auto &h : p.holdings) {
    if (h.coin == coin)
      return &h;
    if (!h.coin && !empty)
      empty = &h;
  }
  return empty;
}
uint16_t owned(const Player &p, unsigned coin) {
  for (auto &h : p.holdings)
    if (h.coin == coin)
      return h.quantity;
  return 0;
}
void Market::reset(uint32_t seed) {
  *this = Market{};
  id = seed ? seed : 1;
  dirty = true;
}
int Market::find(Mac m) const {
  for (int i = 0; i < players; i++)
    if (p[i].mac == m)
      return i;
  return -1;
}
int Market::join(Mac m) {
  int i = find(m);
  if (i >= 0)
    return i;
  if (players == MaxPlayers)
    return -1;
  p[players] = Player{};
  p[players].mac = m;
  dirty = true;
  revision++;
  return players++;
}
static bool symbol_ok(const char *s) {
  size_t n = 0;
  while (n < 6 && s[n])
    n++;
  if (n < 2 || n > 5 || s[0] < 'A' || s[0] > 'Z')
    return false;
  for (size_t i = 1; i < n; i++)
    if ((s[i] < 'A' || s[i] > 'Z') && (s[i] < '0' || s[i] > '9'))
      return false;
  return true;
}
Result Market::request(uint8_t pid, const Request &r, uint64_t now,
                       uint32_t entropy) {
  if (pid >= players)
    return {Error::Player, 0, 0};
  auto &a = p[pid];
  auto hash = request_hash(r);
  if (r.seq == a.lastSeq && r.seq)
    return hash == a.lastHash ? a.last : Result{Error::Sequence, 0, 0};
  if (r.seq != a.next)
    return {Error::Sequence, 0, 0};
  if (r.seq == UINT32_MAX || revision >= UINT32_MAX - 1)
    return {Error::Exhausted, 0, 0};
  Result result{};
  auto fail = [&](Error e) { result.code = e; };
  Coin *coin = r.coin && r.coin <= coins ? &c[r.coin - 1] : nullptr;
  switch (r.op) {
  case Op::Create:
    if (!symbol_ok(r.symbol)) {
      fail(Error::Symbol);
      break;
    }
    if (coins == MaxCoins) {
      fail(Error::Full);
      break;
    }
    if (epoch < a.cooldown) {
      fail(Error::Cooldown);
      break;
    }
    if (a.balance < 2000) {
      fail(Error::Funds);
      break;
    }
    for (unsigned i = 0; i < coins; i++) {
      if (c[i].creator == pid && !c[i].rugged)
        fail(Error::Active);
      if (!std::strcmp(c[i].symbol, r.symbol))
        fail(Error::Symbol);
    }
    if (result.code == Error::Ok) {
      auto &v = c[coins++];
      v = Coin{};
      std::memcpy(v.symbol, r.symbol, 6);
      v.creator = pid;
      v.created = epoch;
      a.balance -= 2000;
      result.coin = coins;
    }
    break;
  case Op::Buy:
  case Op::Sell: {
    bool buy = r.op == Op::Buy;
    int n = r.amount;
    if (!coin || (buy && coin->rugged)) {
      fail(Error::Closed);
      break;
    }
    auto *h = holding(a, r.coin);
    int q = h ? h->quantity : 0;
    auto gross = cost(buy ? coin->supply : coin->supply - n, n);
    if (!buy && q < n) {
      fail(Error::Owned);
      break;
    }
    if (gross < 0) {
      fail(Error::Amount);
      break;
    }
    uint32_t fee = coin->rugged ? 0 : (gross + 49) / 50;
    if (uint64_t(coin->creatorFees) + coin->communityFees + fee > WalletCap) {
      fail(Error::Amount);
      break;
    }
    if (buy && !h) {
      fail(Error::Holdings);
      break;
    }
    if (buy && a.balance < gross + fee) {
      fail(Error::Funds);
      break;
    }
    if (!buy &&
        (coin->reserve < gross || a.balance + gross - fee > WalletCap)) {
      fail(Error::Amount);
      break;
    }
    if (buy) {
      a.balance -= gross + fee;
      coin->reserve += gross;
      coin->supply += n;
      if (!q) {
        h->since = epoch;
        h->eligible = 0;
      }
      h->coin = r.coin;
      h->quantity = q + n;
      if (!(coin->seen & (1 << pid))) {
        coin->seen |= 1 << pid;
        coin->meme = std::min(1000, int(coin->meme) + 10);
      }
    } else {
      a.balance += gross - fee;
      coin->reserve -= gross;
      coin->supply -= n;
      h->quantity = q - n;
      h->eligible = std::min(h->eligible, h->quantity);
      h->since = epoch;
      if (!h->quantity)
        *h = Holding{};
    }
    coin->creatorFees += fee / 2;
    coin->communityFees += fee - fee / 2;
    result.coin = r.coin;
    break;
  }
  case Op::Withdraw: {
    int percent = r.amount;
    if (!coin || coin->rugged) {
      fail(Error::Closed);
      break;
    }
    if (coin->creator != pid) {
      fail(Error::Creator);
      break;
    }
    if (percent != 10 && percent != 25 && percent != 50 && percent != 100) {
      fail(Error::Amount);
      break;
    }
    uint32_t amount = percent == 100
                          ? coin->creatorFees + coin->communityFees
                          : uint64_t(coin->creatorFees) * percent / 100;
    if (uint64_t(a.balance) + amount > WalletCap) {
      fail(Error::Amount);
      break;
    }
    a.balance += amount;
    if (percent == 100) {
      coin->creatorFees = coin->communityFees = 0;
      coin->rugged = true;
      coin->meme = 0;
      a.rep = a.rep > 35 ? a.rep - 35 : 0;
      a.rugs++;
      a.cooldown = std::min(65535, int(epoch) + 3);
    } else
      coin->creatorFees -= amount;
    result.coin = r.coin;
    break;
  }
  case Op::Ticket:
    if (a.gameAt && now - a.gameAt < 20000) {
      fail(Error::Wait);
      break;
    }
    if (a.budget >= 6000) {
      fail(Error::Budget);
      break;
    }
    a.ticket = entropy ? entropy : 1;
    a.gameAt = now ? now : 1;
    result.ticket = a.ticket;
    break;
  case Op::Claim: {
    uint64_t elapsed = now >= a.gameAt ? now - a.gameAt : 0;
    unsigned score = 0, total = 0;
    if (!a.ticket || elapsed < 4000 || elapsed > 45000)
      fail(Error::Proof);
    for (unsigned i = 0; i < 8; i++) {
      unsigned delay;
      auto key = command(a.ticket, i + 1, delay);
      unsigned dt = r.proof[i + 8] * 20;
      auto pressed = r.proof[i];
      if (pressed > 3 || dt < 100 || dt > 2400)
        fail(Error::Proof);
      if (pressed == key && dt >= 100 && dt <= 1800)
        score += 250;
      total += delay + dt;
    }
    if (total > elapsed + 1000)
      fail(Error::Proof);
    if (result.code == Error::Ok) {
      uint32_t reward =
          std::min<uint32_t>({score, 6000u - a.budget, WalletCap - a.balance});
      a.balance += reward;
      a.budget += reward;
    }
    a.ticket = 0;
    break;
  }
  default:
    fail(Error::Amount);
    break;
  }
  a.lastSeq = r.seq;
  a.next = r.seq + 1;
  a.lastHash = hash;
  a.last = result;
  dirty = true;
  revision++;
  return result;
}
void Market::advance_epoch() {
  if (epoch >= 65000)
    return;
  epoch++;
  for (unsigned j = 0; j < coins; j++) {
    auto &coin = c[j];
    uint32_t weight = 0;
    for (unsigned i = 0; i < players; i++)
      for (auto &h : p[i].holdings)
        if (h.coin == j + 1)
          weight +=
              h.eligible *
              std::min(125, 100 + std::max(0, int(epoch) - h.since - 2) * 10);
    uint32_t pool = coin.rugged ? 0 : coin.communityFees / 2;
    for (unsigned i = 0; i < players; i++)
      for (auto &h : p[i].holdings)
        if (h.coin == j + 1) {
          uint32_t reward =
              weight
                  ? uint64_t(pool) * h.eligible *
                        std::min(125,
                                 100 + std::max(0, int(epoch) - h.since - 2) *
                                           10) /
                        weight
                  : 0;
          reward = std::min(reward, WalletCap - p[i].balance);
          p[i].balance += reward;
          coin.communityFees -= reward;
          h.eligible = h.quantity;
        }
    coin.meme = coin.meme * 95 / 100;
    if (!coin.rugged && coin.supply && (epoch - coin.created) % 3 == 0)
      p[coin.creator].rep = std::min(100, int(p[coin.creator].rep) + 1);
  }
  for (unsigned i = 0; i < players; i++) {
    p[i].budget = 0;
    p[i].ticket = 0;
    p[i].gameAt = 0;
  }
  dirty = true;
  revision++;
}
bool Market::valid() const {
  if (!id || !epoch || epoch > 65000 || players > MaxPlayers ||
      coins > MaxCoins)
    return false;
  for (unsigned i = 0; i < players; i++) {
    auto &a = p[i];
    if (a.balance > WalletCap || !a.next || a.rep > 100 || a.budget > 6000 ||
        unsigned(a.last.code) > 16 || a.last.coin > coins ||
        a.lastSeq >= a.next)
      return false;
    for (unsigned k = 0; k < i; k++)
      if (p[k].mac == a.mac)
        return false;
    for (unsigned j = 0; j < MaxHoldings; j++) {
      auto &h = a.holdings[j];
      if (h.coin > coins || h.quantity > 10000 || h.eligible > h.quantity ||
          h.since > epoch || (!h.coin && h.quantity) || (h.coin && !h.quantity))
        return false;
      for (unsigned k = 0; k < j; k++)
        if (h.coin && a.holdings[k].coin == h.coin)
          return false;
    }
  }
  for (unsigned j = 0; j < coins; j++) {
    auto &coin = c[j];
    if (!symbol_ok(coin.symbol) || coin.creator >= players ||
        coin.supply > 10000 || coin.meme > 1000 || coin.created > epoch ||
        coin.reserve != (coin.supply ? cost(0, coin.supply) : 0) ||
        uint64_t(coin.creatorFees) + coin.communityFees > WalletCap)
      return false;
    for (unsigned k = 0; k < j; ++k)
      if (!std::strcmp(coin.symbol, c[k].symbol))
        return false;
    unsigned supply = 0;
    for (unsigned i = 0; i < players; i++)
      supply += owned(p[i], j + 1);
    if (supply != coin.supply)
      return false;
  }
  return true;
}
size_t encode(const Market &m, uint8_t *b, size_t cap) {
  Writer w{b, cap};
  w.u32(0x324d424e);
  w.u32(m.id);
  w.u32(m.revision);
  w.u16(m.epoch);
  w.u8(m.players);
  w.u8(m.coins);
  for (unsigned i = 0; i < m.players; i++) {
    auto &p = m.p[i];
    w.bytes(p.mac.b, 6);
    w.u32(p.balance);
    w.u32(p.next);
    w.u32(p.lastSeq);
    w.u32(p.lastHash);
    w.u8(p.rep);
    w.u16(p.rugs);
    w.u16(p.cooldown);
    w.u16(p.budget);
    w.u8(uint8_t(p.last.code));
    w.u8(p.last.coin);
    w.u32(p.last.ticket);
    w.u32(p.ticket);
    for (auto &h : p.holdings) {
      w.u8(h.coin);
      w.u16(h.quantity);
      w.u16(h.eligible);
      w.u16(h.since);
    }
  }
  for (unsigned i = 0; i < m.coins; i++) {
    auto &c = m.c[i];
    w.bytes(c.symbol, 6);
    w.u8(c.creator);
    w.u8(c.rugged);
    w.u16(c.supply);
    w.u16(c.meme);
    w.u16(c.seen);
    w.u16(c.created);
    w.u32(c.reserve);
    w.u32(c.creatorFees);
    w.u32(c.communityFees);
  }
  auto crc = crc32(b, w.pos);
  w.u32(crc);
  return w.ok ? w.pos : 0;
}
bool decode(Market &out, const uint8_t *b, size_t n, bool restore) {
  if (n < 20 || n > MaxSave)
    return false;
  Reader tail{b + n - 4, 4};
  if (crc32(b, n - 4) != tail.u32())
    return false;
  Reader r{b, n - 4};
  if (r.u32() != 0x324d424e)
    return false;
  Market m{};
  m.id = r.u32();
  m.revision = r.u32();
  m.epoch = r.u16();
  m.players = r.u8();
  m.coins = r.u8();
  if (m.players > MaxPlayers || m.coins > MaxCoins)
    return false;
  for (unsigned i = 0; i < m.players; i++) {
    auto &p = m.p[i];
    r.bytes(p.mac.b, 6);
    p.balance = r.u32();
    p.next = r.u32();
    p.lastSeq = r.u32();
    p.lastHash = r.u32();
    p.rep = r.u8();
    p.rugs = r.u16();
    p.cooldown = r.u16();
    p.budget = r.u16();
    p.last.code = Error(r.u8());
    p.last.coin = r.u8();
    p.last.ticket = r.u32();
    p.ticket = r.u32();
    for (auto &h : p.holdings) {
      h.coin = r.u8();
      h.quantity = r.u16();
      h.eligible = r.u16();
      h.since = r.u16();
    }
    if (restore) {
      p.ticket = 0;
      p.last.ticket = 0;
    }
  }
  for (unsigned i = 0; i < m.coins; i++) {
    auto &c = m.c[i];
    r.bytes(c.symbol, 6);
    c.creator = r.u8();
    auto rugged = r.u8();
    if (rugged > 1)
      return false;
    c.rugged = rugged;
    c.supply = r.u16();
    c.meme = r.u16();
    c.seen = r.u16();
    c.created = r.u16();
    c.reserve = r.u32();
    c.creatorFees = r.u32();
    c.communityFees = r.u32();
  }
  if (!r.ok || r.pos != n - 4 || !m.valid())
    return false;
  out = m;
  return true;
}
} // namespace bm

File: app/badge_market/market.lua
local T=require('copy')
-- Canonical authority. Packed records keep inactive players/coins off Lua tables.
local C = require('codec')
local M = { MAX_PLAYERS=12, MAX_COINS=16, HOLDINGS=4 }
local pack, unpack = string.pack, string.unpack
local PF = '<c6I4I2I2I2I2I1I1I2I2I4I4I4I4'
-- MAC,balance,nextSeq,rep,rugs,cooldown,lastCode,lastCoin,lastSeq,budget,hash,ticket,lastGame
local CF = '<I1I2I4I4I4I2I2I2I1c5'
-- creator,supply,reserve,creatorFee,communityFee,meme,seenBits,createdEpoch,rug,symbol
local HF = '<I1I2I2I2' -- coin,quantity,eligible,startEpoch
local PS = string.packsize(PF)
local CS = string.packsize(CF)
local empty = pack(HF,0,0,0,0):rep(4)
function M.error(code) return T(code>=0 and code<=16 and (192+code) or 177) end
local function player(m, id)
  local s=m.p[id]; if not s then return nil end
  local t={unpack(PF,s)}; t[15]=s:sub(PS+1); return t
end
local function putp(m,id,p)
  m.p[id]=pack(PF,table.unpack(p,1,14))..p[15]; m.dirty=true
end
local function coin(m,id)
  if not m.c[id] then return nil end
  return {unpack(CF,m.c[id])}
end
local function putc(m,id,c) m.c[id]=pack(CF,table.unpack(c,1,10)); m.dirty=true end
local function holding_data(data,id,offset)
  local free
  for i=1,4 do
    local a,b,c,d=unpack(HF,data,offset+(i-1)*7+1)
    if a==id then return i,b,c,d end
    if a==0 then free=i end
  end
  return free,0,0,0
end
local function holding(p,id) return holding_data(p[15],id,0) end
local function puth(p,i,id,q,e,start)
  local pos=(i-1)*7
  p[15]=p[15]:sub(1,pos)..pack(HF,q>0 and id or 0,q,e,start)..p[15]:sub(pos+8)
end
function M.new(id) return {id=id,epoch=1,p={},c={},dirty=true,gen=0} end
function M.join(m,mac)
  if #mac~=6 then return nil end
  for i=1,#m.p do if m.p[i]:sub(1,6)==mac then return i end end
  if #m.p>=M.MAX_PLAYERS then return nil end
  local id=#m.p+1
  putp(m,id,{mac,25000,1,75,0,0,0,0,0,0,0,0,0,0,empty})
  return id
end
function M.wallet(m,id)
  local p=player(m,id); if not p then return end
  return p[2],p[3],p[4],p[5],p[15],p[10]
end
function M.info(m,id,who)
  if not m.c[id] then return end
  local creator,supply,_,cf,hf,meme,_,_,rug,symbol=unpack(CF,m.c[id])
  local q=0
  if m.p[who] then local _; _,q=holding_data(m.p[who],id,PS) end
  local holders=0
  for i=1,#m.p do local _,n=holding_data(m.p[i],id,PS); if n>0 then holders=holders+1 end end
  return symbol:gsub('%z',''),supply,creator,cf+hf,meme,holders,rug,q
end
function M.cost(q,n)
  if math.type(q)~='integer' or math.type(n)~='integer' or q<0 or n<1 or q+n>10000 then return nil end
  return n*100 + (5*n*(2*q+n-1))//2
end
-- Sequential mutations: duplicates replay their original result, never mutate twice.
function M.request(m,id,seq,op,a,b,now)
  local p=player(m,id); if not p then return 2,0 end
  local hash=C.crc(string.char(op)..pack('<I2',a or 0)..(b or ''))
  if seq==p[9] then
    if hash==p[11] then return p[7],p[8],p[12] end
    return 3,0
  end
  if seq~=p[3] then return 3,0 end
  if seq>=65535 then return 16,0 end
  local code,which=0,0
  local c=coin(m,a)
  if op==1 then -- create, a is preset icon, b is symbol
    if not b or #b<2 or #b>5 or not b:match('^[A-Z][A-Z0-9]+$') then code=12
    elseif #m.c>=M.MAX_COINS then code=1
    elseif m.epoch<p[6] then code=10
    elseif p[2]<2000 then code=4
    else
      for i=1,#m.c do
        local other=coin(m,i)
        if other[1]==id and other[9]==0 then code=11 end
        if other[10]:gsub('%z','')==b then code=12 end
      end
      if code==0 then
        which=#m.c+1; p[2]=p[2]-2000
        putc(m,which,{id,0,0,0,0,0,0,m.epoch,0,b})
      end
    end
  elseif op==2 or op==3 then -- buy or sell, amount is a 16-bit LE payload
    local n=b and #b==2 and unpack('<I2',b) or 0
    if not c or (op==2 and c[9]>0) then code=8
    elseif n<1 or n>10000 then code=5
    else
      local slot,q,e,start=holding(p,a)
      local gross=M.cost(op==2 and c[2] or c[2]-n,n)
      if op==3 and q<n then code=7
      elseif not gross then code=5
      else
        local fee=c[9]>0 and 0 or (gross+49)//50
        if op==2 and not slot then code=6
        elseif op==2 and p[2]<gross+fee then code=4
        elseif op==3 and q<n then code=7
        elseif op==3 and c[3]<gross then code=8
        elseif op==3 and p[2]+gross-fee>1000000000 then code=5
        else
          local sign=op==2 and 1 or -1
          p[2]=p[2]-sign*gross-fee; c[2]=c[2]+sign*n; c[3]=c[3]+sign*gross
          if c[9]==0 then c[4]=c[4]+fee//2; c[5]=c[5]+fee-fee//2
          else p[2]=p[2]+fee end -- exits after rug are fee-free
          if op==2 then
            if q==0 then start=m.epoch; e=0 end
            if c[7] & (1<<(id-1))==0 then c[7]=c[7] | (1<<(id-1)); c[6]=math.min(1000,c[6]+10) end
          else e=math.min(e,q-n); start=m.epoch end
          puth(p,slot,a,q+sign*n,e,start); putc(m,a,c); which=a
        end
      end
    end
  elseif op==4 then -- creator withdrawal or full rug
    local percent=b and b:byte(1) or 0
    if not c or c[9]>0 then code=8
    elseif c[1]~=id then code=9
    elseif percent~=10 and percent~=25 and percent~=50 and percent~=100 then code=5
    elseif p[2]+(percent==100 and c[4]+c[5] or c[4]*percent//100)>1000000000 then code=5
    else
      local amount=c[4]*percent//100
      if percent==100 then
        amount=c[4]+c[5]; c[5]=0; c[9]=1; c[6]=0
        p[4]=math.max(0,p[4]-35); p[5]=p[5]+1; p[6]=m.epoch+3
      end
      c[4]=c[4]-(percent==100 and c[4] or amount); p[2]=p[2]+amount
      putc(m,a,c); which=a
    end
  elseif op==5 then -- host issues unpredictable one-use reaction ticket
    if p[13]>0 and now-p[13]<20000 then code=13
    elseif p[10]>=6000 then code=15
    else p[12]=math.max(1,C.seed(m.id*65536+now+id+seq)); p[13]=math.max(1,now); p[14]=now end
  elseif op==6 then -- 8 response keys plus 8 times in 20 ms units
    if p[12]==0 or not b or #b~=16 or now-p[14]>45000 or now-p[14]<4000 then code=14
    else
      local score,total=0,0
      for i=1,8 do
        local key,delay=C.command(p[12],i); local press,dt=b:byte(i),b:byte(i+8)*20
        if press>3 or dt<100 or dt>2400 then code=14 end
        if press==key and dt>=100 and dt<=1800 then score=score+250 end
        total=total+delay+dt
      end
      if total>now-p[14]+1000 then code=14 end
      if code==0 then local reward=math.min(score,6000-p[10],1000000000-p[2]); p[2]=p[2]+reward; p[10]=p[10]+reward end
    end
    p[12]=0
  else code=5 end
  p[3]=seq+1; p[7]=code; p[8]=which; p[9]=seq; p[11]=hash
  putp(m,id,p)
  return code,which,p[12]
end
function M.epoch(m)
  if m.epoch>=65000 then return false end
  m.epoch=m.epoch+1
  for id=1,#m.c do
    local c=coin(m,id); local weight=0
    for i=1,#m.p do
      local _,_,e,start=holding_data(m.p[i],id,PS)
      weight=weight+e*math.min(125,100+math.max(0,m.epoch-start-2)*10)
    end
    local pool=c[9]==0 and c[5]//2 or 0
    for i=1,#m.p do
      local slot,q,e,start=holding_data(m.p[i],id,PS)
      if q>0 then
        local p=player(m,i)
        local reward=weight>0 and pool*e*math.min(125,100+math.max(0,m.epoch-start-2)*10)//weight or 0
        reward=math.min(reward,1000000000-p[2]); p[2]=p[2]+reward; c[5]=c[5]-reward
        puth(p,slot,id,q,q,start); putp(m,i,p)
      end
    end
    c[6]=c[6]*95//100; putc(m,id,c)
    if c[9]==0 and c[2]>0 and (m.epoch-c[8])%3==0 then
      local owner=player(m,c[1]); owner[4]=math.min(100,owner[4]+1); putp(m,c[1],owner)
    end
  end
  for i=1,#m.p do local p=player(m,i); p[10]=0; p[12]=0; p[13]=0; putp(m,i,p) end
  m.dirty=true; return true
end
function M.encode(m)
  return C.wrap(pack('<c4I4I2I2I1I1','BMS1',m.gen+1,m.id,m.epoch,#m.p,#m.c)..table.concat(m.p)..table.concat(m.c))
end
function M.decode(s)
  if type(s)~='string' or #s<18 or #s>4096 or s:sub(1,4)~='BMS1' or not C.valid(s) then return nil end
  local _,gen,id,epoch,np,nc,pos=unpack('<c4I4I2I2I1I1',s)
  if np>12 or nc>16 or epoch<1 or epoch>65000 or #s~=18+np*(PS+28)+nc*CS then return nil end
  local m=M.new(id); m.gen=gen; m.epoch=epoch
  for i=1,np do
    m.p[i]=s:sub(pos,pos+PS+27); pos=pos+PS+28
    local p=player(m,i)
    if p[2]>1000000000 or p[3]==0 then return nil end
    for j=1,4 do local cid,q,e=unpack(HF,p[15],(j-1)*7+1); if cid>nc or q>10000 or e>q then return nil end end
    p[12]=0; p[13]=0; p[14]=0; putp(m,i,p)
  end
  for i=1,nc do
    m.c[i]=s:sub(pos,pos+CS-1); pos=pos+CS
    local c=coin(m,i)
    if c[1]<1 or c[1]>np or c[2]>10000 or c[9]>1 or c[3]~=(c[2]>0 and M.cost(0,c[2]) or 0) then return nil end
    local supply=0
    for j=1,np do local _,q=holding(player(m,j),i); supply=supply+q end
    if supply~=c[2] then return nil end
  end
  m.dirty=false; return m
end
return M
